File: test/fixtures/tool_abuse.py
# Smoke-test fixture: tool_abuse
#
# A LangChain agent is given a `run_shell` tool that executes arbitrary shell
# commands chosen by the LLM. There's no per-tool authorization gate, no
# input validation, no allow-list of safe commands, and no confirmation
# step. An attacker who can influence the LLM's input (via prompt injection
# or a poisoned RAG context) can ask the agent to run any shell command —
# including reading files, exfiltrating data, or pivoting to the host.
import subprocess

from langchain.agents import AgentExecutor, create_react_agent
from langchain.tools import StructuredTool
from langchain_openai import ChatOpenAI


def run_shell(command: str) -> str:
    """Run a shell command and return its output."""
    # Tool abuse: no validation, no allow-list, no auth check, shell=True.
    # The LLM can call this with any string and the process executes it.
    result = subprocess.run(command, shell=True, capture_output=True, text=True)
    return result.stdout + result.stderr


def read_file(path: str) -> str:
    """Read the contents of an arbitrary file from disk."""
    # Tool abuse: arbitrary file read by LLM choice.
    with open(path) as f:
        return f.read()


tools = [
    StructuredTool.from_function(run_shell),
    StructuredTool.from_function(read_file),
]

llm = ChatOpenAI(model="gpt-4o-mini", api_key="sk-PLACEHOLDER")
agent = create_react_agent(llm, tools, prompt="You are a helpful assistant.")
executor = AgentExecutor(agent=agent, tools=tools)

# Anywhere a user-controlled string flows into `executor.invoke({"input": ...})`,
# the LLM can choose to call run_shell or read_file with attacker-chosen args.

File: test/fixtures/tool_abuse.ts
// Smoke-test fixture: tool_abuse
//
// An MCP server exposes a `write_file` tool to LLM clients with no
// authorization check, no path validation (so the LLM can write outside
// the working directory), and no confirmation step. Any client that
// connects to this MCP server can ask the assistant to overwrite any file
// the server process has permission to touch.

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { writeFileSync } from "fs";

const server = new McpServer({ name: "file-tools", version: "1.0.0" });

// Tool abuse: write_file accepts an arbitrary path and content. No
// allow-list of writable directories, no canonicalisation against a
// project root, no auth check on the calling LLM session.
server.tool(
  "write_file",
  {
    path: z.string().describe("Absolute or relative path to write to"),
    content: z.string().describe("Content to write to the file"),
  },
  async ({ path, content }) => {
    writeFileSync(path, content);
    return {
      content: [{ type: "text", text: `Wrote ${content.length} bytes to ${path}` }],
    };
  }
);

// Tool abuse: delete_file with the same pattern — arbitrary FS write.
server.tool(
  "delete_file",
  { path: z.string() },
  async ({ path }) => {
    const { unlinkSync } = await import("fs");
    unlinkSync(path);
    return { content: [{ type: "text", text: `Deleted ${path}` }] };
  }
);

await server.connect();
